### PR TITLE
Add json output format to several commands

### PR DESCRIPTION
**Pull request by ulyssessouza**
**What I did**
Add json output format to several commands

- [x] docker context ls - To make the `--json` hidden and implement `--format json`
- [x] docker ps
- [x] docker compose ls
- [x] docker compose ps
- [x] docker secret ls
- [x] docker volume ls
- [x] docker version - To clarify retro compatibility issues
- [x] Add tests

**Related issue**
Resolves https://github.com/docker/compose-cli/issues/607

**(not mandatory) A picture of a cute animal, if possible in relation with what you did**
format

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -62,5 +62,6 @@ require (
 	google.golang.org/protobuf v1.25.0
 	gopkg.in/check.v1 v1.0.0-20190902080502-41f04d3bba15 // indirect
 	gopkg.in/ini.v1 v1.61.0
+	gotest.tools v2.2.0+incompatible
```
Comment (ulyssessouza) [inline go.mod:65]: This is the result of a `go mod tidy`

File: cli/cmd/version.go [hunk 2]
```diff
@@ -18,43 +18,94 @@ package cmd
...
 	// Still, technically the [] byte versionResult could be nil, just let the original command display what it has to display
 	if versionResult == nil {
 		mobycli.Exec(cmd.Root())
-		return nil
+		return ""
+	}
+	return string(versionResult)
+}
+
+func fixedPrettyArgs(oArgs []string) []string {
```
Comment (ulyssessouza) [inline cli/cmd/version.go:80]: Not very happy with this, but works fine

File: cli/cmd/version.go [hunk 3]
```diff
@@ -18,43 +18,94 @@ package cmd
...
+			(strings.ToLower(oArgs[i+1]) == formatter.PRETTY || oArgs[i+1] == "") {
+			i++
+			continue
+		}
+		args = append(args, oArgs[i])
+	}
+	return args
+}
+
+func fixedJsonArgs(oArgs []string) []string {
```
Comment (ulyssessouza) [inline cli/cmd/version.go:94]: Not very happy with this, but works fine

File: cli/cmd/context/ls.go [hunk 4]
```diff
@@ -93,35 +95,41 @@ func runList(cmd *cobra.Command, opts lsOpts) error {
...
+	return printContextLsFormatted(opts.format, currentContext, os.Stdout, contexts)
+}
 
-	for _, c := range contexts {
-		contextName := c.Name
-		if c.Name == currentContext {
-			contextName += " *"
+func printContextLsFormatted(format string, currContext string, out io.Writer, contexts []*store.DockerContext) error {
+	var err error
+	switch strings.ToLower(format) {
```
Comment (rumpl) [inline cli/cmd/context/ls.go:106]: What about
```golang
err = formatter.Print(context, format, func(w io.Writer) {
	for _, c := range contexts {
		contextName := c.Name
		if c.Name == currContext {
			contextName += " *"
		}
		fmt.Fprintf(w, "%s\t%s\t%s\t%s\t%s\t%s\n",
			contextName,
			c.Type(),
			c.Metadata.Description,
			getEndpoint("docker", c.Endpoints),
			getEndpoint("kubernetes", c.Endpoints),
			c.Metadata.StackOrchestrator)
	}
} , "NAME", "TYPE", "DESCRIPTION", "DOCKER ENDPOINT", "KUBERNETES ENDPOINT", "ORCHESTRATOR")
```

And we let the formatter switch the format and call the `ToStandardJSON` when needed?
This way this can be reused everywhere.

**Review by gtardif — COMMENTED** [review]
We should return `[]` and not null when there are no elements : 
```
$ ./bin/docker --context aci  ps --format json
null
```

File: cli/cmd/compose/ps.go [hunk 5]
```diff
@@ -59,17 +61,26 @@ func runPs(ctx context.Context, opts composeOptions) error {
...
+	var err error
+	switch strings.ToLower(format) {
+	case formatter.PRETTY, "":
+		err = formatter.PrintPrettySection(out, func(w io.Writer) {
+			for _, service := range serviceList {
+				fmt.Fprintf(w, "%s\t%s\t%d/%d\t%s\n", service.ID, service.Name, service.Replicas, service.Desired, strings.Join(service.Ports, ", "))
+			}
+		}, "ID", "NAME", "REPLICAS", "PORTS")
+	case formatter.JSON:
+		outJSON, err := formatter.ToStandardJSON(serviceList)
```
Comment (gtardif) [inline cli/cmd/compose/ps.go:77]: Same question here, although it does not expose too many additional things. At least we should set json attribut names in the go struct, current field case is mainly driven by public/private fields and go linter, that should not drive the json output

File: cli/cmd/compose/testdata/compose-list-out-json.golden [hunk 6]
```diff
@@ -0,0 +1,7 @@
+[
+    {
+        "ID": "123",
+        "Name": "myName123",
+        "Status": "Running"
+    }
+]
```
Comment (gtardif) [inline cli/cmd/compose/testdata/compose-list-out-json.golden:7]: nit

File: cli/cmd/context/ls.go [hunk 7]
```diff
@@ -58,7 +59,8 @@ func listCommand() *cobra.Command {
 	}
 	cmd.Flags().BoolVarP(&opts.quiet, "quiet", "q", false, "Only show context names")
 	cmd.Flags().BoolVar(&opts.json, "json", false, "Format output as JSON")
-	cmd.Flags().StringVar(&opts.format, "format", "", "Format output as JSON")
+	cmd.Flags().StringVar(&opts.format, "format", "", "Format the output. Values: [pretty | json | go template]. (Default: pretty)")
```
Comment (gtardif) [inline cli/cmd/context/ls.go:62]: We support go templating here only for backward compatibility, not sure we should mention it in the doc

File: cli/cmd/context/ls.go (hunk 4 above)
Comment (gtardif) [reply cli/cmd/context/ls.go:106]: +1. This goes with my concerns above regarding exposing our go structs and not returning the same information depending on format json or pretty. Using the formatter to output both json and pretty enforces better choices there

File: cli/cmd/ps.go [hunk 8]
```diff
@@ -59,50 +54,69 @@ func PsCommand() *cobra.Command {
 	cmd.Flags().BoolVarP(&opts.quiet, "quiet", "q", false, "Only display IDs")
 	cmd.Flags().BoolVarP(&opts.all, "all", "a", false, "Show all containers (default shows just running)")
 	cmd.Flags().BoolVar(&opts.json, "json", false, "Format output as JSON")
+	cmd.Flags().StringVar(&opts.format, "format", "", "Format the output. Values: [pretty | json | go template]. (Default: pretty)")
```
Comment (gtardif) [inline cli/cmd/ps.go:57]: Same here, we support got template for backward compatibility reason here, not sure we want to document it for the long run

File: cli/cmd/ps.go [hunk 9]
```diff
@@ -59,50 +54,69 @@ func PsCommand() *cobra.Command {
...
+	_ = cmd.Flags().MarkHidden("json")
```
Comment (gtardif) [inline cli/cmd/ps.go:58]: Maybe a comment mentionning this is used in VSCode Docker extension for the moment

File: cli/cmd/ps.go [hunk 10]
```diff
@@ -59,50 +54,69 @@ func PsCommand() *cobra.Command {
...
+	switch strings.ToLower(format) {
+	case formatter2.PRETTY, "":
+		err = formatter2.PrintPrettySection(out, func(w io.Writer) {
+			for _, c := range containers {
+				fmt.Fprintf(w, "%s\t%s\t%s\t%s\t%s\n", c.ID, c.Image, c.Command, c.Status,
+					strings.Join(formatter.PortsToStrings(c.Ports, fqdn(c)), ", "))
+			}
+		}, "CONTAINER ID", "IMAGE", "COMMAND", "STATUS", "PORTS")
+	case formatter2.JSON:
+		out, err := formatter2.ToStandardJSON(containers)
```
Comment (gtardif) [inline cli/cmd/ps.go:110]: Same here

File: cli/cmd/version.go [hunk 11]
```diff
@@ -18,43 +18,94 @@ package cmd
...
+	var versionString string
+	format := strings.TrimSpace(cmd.Flag(formatOpt).Value.String())
 	displayedVersion := strings.TrimPrefix(version, "v")
-	versionResult, _ := mobycli.ExecSilent(cmd.Context())
+	// Replace is preferred in this case to keep the order.
+	switch format {
+	case formatter.PRETTY, "":
+		versionString = strings.Replace(getOutFromMoby(cmd, fixedPrettyArgs(os.Args[1:])...),
+			"\n Version:", "\n Cloud integration:  "+displayedVersion+"\n Version:", 1)
+	case formatter.JSON, "{{json .}}", "{{json . }}", "{{ json .}}", "{{ json . }}": // Try to catch full JSON formats
```
Comment (gtardif) [inline cli/cmd/version.go:62]: easier if you replace space first ? (`replace(" ", "")`)

Comment (gtardif) [reply cli/cmd/version.go:62]: and toLowerCase

File: cli/mobycli/exec.go [hunk 12]
```diff
@@ -112,7 +112,10 @@ func IsDefaultContextCommand(dockerCommand string) bool {
 }
 
 // ExecSilent executes a command and do redirect output to stdOut, return output
-func ExecSilent(ctx context.Context) ([]byte, error) {
-	cmd := exec.CommandContext(ctx, ComDockerCli, os.Args[1:]...)
+func ExecSilent(ctx context.Context, args ...string) ([]byte, error) {
+	if len(args) == 0 {
```
Comment (gtardif) [inline cli/mobycli/exec.go:116]: How was this change required ? maybe we should change the caller to make sure args include the `docker` command

File: cli/cmd/compose/testdata/compose-list-out-json.golden (hunk 6 above)
Comment (ulyssessouza) [reply cli/cmd/compose/testdata/compose-list-out-json.golden:7]: That's the good one. It's a golden file

File: cli/mobycli/exec.go (hunk 12 above)
Comment (ulyssessouza) [reply cli/mobycli/exec.go:116]: This is used on `getOutFromMoby` when shelling out to moby. The idea is to hack the `format` option before passing it to moby, since it should treat the cases of `pretty` and `json`. Both are inexistent on moby, so I do a little trick to translate it.

File: cli/cmd/context/ls.go (hunk 4 above)
Comment (ulyssessouza) [reply cli/cmd/context/ls.go]: Done

File: cli/cmd/ps.go (hunk 9 above)
Comment (ulyssessouza) [reply cli/cmd/ps.go:57]: Done

File: cli/cmd/version.go (hunk 11 above)
Comment (ulyssessouza) [reply cli/cmd/version.go]: Done

File: aci/containers.go [hunk 13]
```diff
@@ -53,7 +53,7 @@ func (cs *aciContainerService) List(ctx context.Context, all bool) ([]containers
 	if err != nil {
 		return nil, err
 	}
-	var res []containers.Container
+	res := []containers.Container{}
```
Comment (rumpl) [inline aci/containers.go:56]: Unrelated change is unrelated :D

Comment (ulyssessouza) [reply aci/containers.go:56]: It is related.

The previous results in `nil` in the case of an empty list. The new results in an `empty slice`.

This is related to the `JSON` serialization. It handles differently `nil` and an `empty slice`.

Comment (ulyssessouza) [reply aci/containers.go:56]: Maybe `make([]containers.Cotainer, 0)` could make that more specific.

**Comment by ulyssessouza** [comment]
@gtardif @rumpl PTAL

**Review by gtardif — CHANGES_REQUESTED** [review]
Couple of things: 
* `docker ps --format json` does not display ports properly : `,"Ports":["20.56.234.156:80-\u003e80/tcp"]`
* More important (moby regression): version command is not forwarded properly to moby when we don't know the fomat: 
` ./bin/docker  version --format "{{ yaml . }}" does not display anything and exits with 0
The moby command  says `Template parsing error: template: version:1: function "yaml" not defined`
Same with ` $ ./bin/docker  version --format "{{ json .Client }}"`: empty result.
(`context ls` does not seem to have this issue).

**Comment by ulyssessouza** [comment]
@gtardif PTAL

**Review by gtardif — APPROVED** [review]

**Comment by BuonOmo** [comment]
Hey, is it normal that this commit is available in v1.0.X, v2.X, but not versions in between ? :) (source: the tags associated with the commit https://github.com/docker/compose-cli/commit/8961805412ba307255f378680a1af903b19b384c)
